File: src/skill-library/registry.ts
import { Skill } from '../types';

export class SkillRegistry {
  private skills: Map<string, Skill> = new Map();

  register(skill: Skill): void {
    this.skills.set(skill.id, skill);
  }

  unregister(skillId: string): void {
    this.skills.delete(skillId);
  }

  get(skillId: string): Skill | undefined {
    return this.skills.get(skillId);
  }

  getByTag(tag: string): Skill[] {
    return Array.from(this.skills.values()).filter(s => s.tags.includes(tag));
  }

  getAll(): Skill[] {
    return Array.from(this.skills.values());
  }

  search(query: string): Skill[] {
    const q = query.toLowerCase();
    return Array.from(this.skills.values()).filter(
      s => s.name.toLowerCase().includes(q) || s.description.toLowerCase().includes(q)
    );
  }

  listVersions(skillName: string): string[] {
    return Array.from(this.skills.values())
      .filter(s => s.name === skillName)
      .map(s => s.version);
  }

  getLatestVersion(skillName: string): Skill | undefined {
    const versions = Array.from(this.skills.values()).filter(s => s.name === skillName);
    if (versions.length === 0) return undefined;
    return versions.sort((a, b) => b.version.localeCompare(a.version, undefined, { numeric: true }))[0];
  }
}

File: agentic_skill_framework/skill_library/registry.py
from typing import Optional
from .skill_definition import Skill
from ..models import SkillMetadata

class SkillRegistry:
    def __init__(self):
        self._skills: dict[str, Skill] = {}

    def register(self, skill: Skill) -> None:
        self._skills[skill.metadata.name] = skill

    def get(self, name: str) -> Optional[Skill]:
        return self._skills.get(name)

    def list_skills(self) -> list[SkillMetadata]:
        return [s.metadata for s in self._skills.values()]

    def find_by_tag(self, tag: str) -> list[Skill]:
        return [s for s in self._skills.values() if tag in s.metadata.tags]

    def count(self) -> int:
        return len(self._skills)

    def unregister(self, name: str) -> bool:
        if name in self._skills:
            del self._skills[name]
            return True
        return False
